### ISSUE TITLE
Regression with new iteration protocol

### ISSUE THREAD

**Issue by traktofon** [JuliaLang/julia#28695]
While trying to upgrade the package [TensorOperations](https://github.com/Jutho/TensorOperations.jl) for Julia 1.0 I came across a case where switching to the new iterator protocol causes a performance regression. I know that not all optimization for the new protocol are in place yet, but maybe documenting this case can help to recover the former performance. Also I think this is a case where the protocol must be used directly, as explicit access to the state is needed.

The code boils down to:
```julia
D = Dict(:a=>1, :b=>2, :c=>3, :d=>4)

function f1(d)
   it = keys(d)
   res = 0
   state1 = start(it)
   while !done(it,state1)
      s1,ns1 = next(it,state1)
      state2 = ns1
      while !done(it,state2)
         s2,ns2 = next(it,state2)
         res += d[s1]*d[s2]
         state2 = ns2
      end
      state1 = ns1
   end
   res
end

function f2(d)
   it = keys(d)
   res = 0
   elst1 = iterate(it)
   while elst1 != nothing
      s1,ns1 = elst1
      elst2 = iterate(it,ns1)
      while elst2 != nothing
         s2,ns2 = elst2
         res += d[s1]*d[s2]
         elst2 = iterate(it,ns2)
      end
      elst1 = iterate(it,ns1)
   end
   res
end
```

In short the function runs over all pairs of elements of the iterator. The computation of the result `res` here is just a stand-in. `f1` uses the old iteration protocol and `f2` the new one. Timings as follows:

Julia 0.6.4:
```
julia> @btime f1($D)
  251.994 ns (0 allocations: 0 bytes)
35
```
Julia 0.7.0:
```
julia> @btime f2($D)
 4.020 μs (30 allocations: 960 bytes)
35
```
(Benchmarking `f1` on 0.7.0 is meaningless because of the deprecations.)

**Comment by traktofon** [comment]
Ah! Thanks a lot! That solves it.
